### PR TITLE
FastExp

### PR DESCRIPTION
**Pull request by StylishTriangles**

### CODE REVIEW

File: tests/inner_txn_test.py [hunk 1]
```diff
@@ -1,19 +1,7 @@
...
-    CompileOptions,
-    Global,
-    Int,
-    Mode,
-    TealBlock,
-    TealComponent,
-    TxnField,
-    TxnType,
-)
+from pyteal import Addr, App, Bytes, CompileOptions, Global, Int, Mode, TealBlock, TealComponent, TxnField, TxnType
```
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply tests/inner_txn_test.py:4]: It was auto-formatted by black (not sure why it's only happening now). The line length limit is 120 here

File: pytealext/evaluator/evaluator.py [hunk 2]
```diff
@@ -624,8 +625,11 @@ def eval_teal(  # pylint: disable=too-many-locals,too-many-branches,too-many-sta
                 stack.append(context.txn.app_args[arg_index])
             else:
                 raise Exception("Unsupported txna expression")
-        elif op == "byte":
+        elif op == "addr":
             arg = args[0]
+            stack.append(decode_address(arg))
+        elif op == "byte":
+            arg = line[5:]
             if arg[0] == '"' and arg[-1] == '"':
```
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply pytealext/evaluator/evaluator.py:633]: This change is a quick solution to a bug that I found -> string "Hello, World!" (from the new docs) wouldn't be parsed properly

This wouldn't work if the line had a comment at the end

File: pytealext/fastexp.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                x.store(Mul128(base, base)),
+                If(exponent == Int(2)).Then(Return(x.load())),
+                If(exponent % Int(2))
+                .Then(
+                    #
+                    y.store(Itob(base))
+                )
+                .Else(
+                    #
+                    y.store(Itob(Int(1)))
```
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply pytealext/fastexp.py:29]: To force formatting in this way, I think I'll roll it back though, it's probably not as unreadable as I thought

File: pytealext/evaluator/__init__.py [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
-__all__ = ["Panic", "AssertionFailed", "EvalContext", "eval_teal", "INTEGER_SIZE"]
+__all__ = [
+    "Panic",
+    "AssertionFailed",
+    "EvalContext",
+    "eval_teal",
+    "INTEGER_SIZE",
+    "compile_and_run",
+    "summarize_execution",
+    "ExecutionSummary",
```
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply pytealext/evaluator/__init__.py:13]: isort doesn't reorder __all__

File: tests/inner_txn_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (chwilko) [reply tests/inner_txn_test.py:4]: ok

File: pytealext/fastexp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chwilko) [reply pytealext/fastexp.py:29]: ok

File: tests/evaluator_test.py [hunk 5]
```diff
@@ -21,8 +21,7 @@
     compileTeal,
 )
 
-from pytealext.evaluator import EvalContext, Panic, eval_teal
-from tests.helpers import compile_and_run
+from pytealext.evaluator import EvalContext, Panic, compile_and_run, eval_teal
```
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply tests/evaluator_test.py:25]: yup

File: tests/fastexp_test.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    expected = base**exponent
+
+    expr = Seq(result.store(FastExp(base_int, exponent_int)), Int(1))
+
+    stack, slots = compile_and_run(expr)
+
+    actual = int.from_bytes(slots[0], "big")
+
+    assert len(stack) == 1
```
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply tests/fastexp_test.py:40]: yep, it's not needed here, in others, where we used assembly directly, we validate that there is only 1 variable left on stack after all operations

File: pytealext/evaluator/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chwilko) [reply pytealext/evaluator/__init__.py:13]: In this file order is important?

File: pytealext/evaluator/evaluator.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chwilko) [reply pytealext/evaluator/evaluator.py:633]: ok

File: tests/evaluator_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chwilko) [reply tests/evaluator_test.py:25]: ok

File: tests/fastexp_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chwilko) [reply tests/fastexp_test.py]: ok

File: pytealext/evaluator/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (StylishTriangles) [reply pytealext/evaluator/__init__.py:13]: not really, `__all__` seems to only matter when you're doing `import *`

[the remark this answers is not in the document]
Comment (chwilko) [reply pytealext/evaluator/__init__.py:13]: ok it is list xD

**Review by chwilko — APPROVED** [review]

**Pull request merged**